src/Sentry/GlobalSessionManager.cs

```diff
@@ -0,0 +1,218 @@
+
+                    var filePath = Path.Combine(directoryPath, ".installation");
+
+                    // Read installation ID stored in a file
+                    try
+                    {
+                        return File.ReadAllText(filePath);
+                    }
+                    catch (FileNotFoundException)
+                    {
+                        _options.DiagnosticLogger?.LogDebug(
+                            "File containing installation ID does not exist ({0}).",
+                            filePath
+                        );
+                    }
+
+                    // Generate new installation ID and store it in a file
+                    var id = Guid.NewGuid().ToString();
+                    File.WriteAllText(filePath, id);
+
+                    _options.DiagnosticLogger?.LogDebug(
+                        "Saved installation ID '{0}' to file '{1}'.",
+                        id, filePath
+                    );
+
+                    return id;
+                }
+                catch (IOException ex)
+                {
+                    _options.DiagnosticLogger?.LogError(
+                        "Failed to resolve persistent installation ID.",
+                        ex
+                    );
+
+                    return null;
+                }
+            }
+        }
+
+        private string? TryGetHardwareInstallationId()
+        {
+            var installationId = NetworkInterface
+                .GetAllNetworkInterfaces()
+                .Where(nic =>
+                    nic.OperationalStatus == OperationalStatus.Up &&
+                    nic.NetworkInterfaceType != NetworkInterfaceType.Loopback)
+                .Select(nic => nic.GetPhysicalAddress().ToString())
+                .FirstOrDefault();
+
+            if (string.IsNullOrWhiteSpace(installationId))
+            {
+                _options.DiagnosticLogger?.LogError(
+                    "Failed to resolve hardware installation ID."
+                );
+
+                return null;
+            }
+
+            return installationId;
+        }
+
+        private string? TryGetInstallationId()
+        {
+            // Installation ID could have already been resolved by this point
+            if (!string.IsNullOrWhiteSpace(_cachedInstallationId))
+            {
+                return _cachedInstallationId;
+            }
+
+            var id =
+                TryGetPersistentInstallationId() ??
+                TryGetHardwareInstallationId();
+
+            if (!string.IsNullOrWhiteSpace(id))
+            {
+                _options.DiagnosticLogger?.LogDebug(
+                    "Resolved installation ID '{0}'.",
+                    id
+                );
+            }
+            else
+            {
+                _options.DiagnosticLogger?.LogDebug(
+                    "Failed to resolve installation ID."
+                );
+            }
+
+            return _cachedInstallationId = id;
```

Comment:
ON a second thought I think this installation id should be a different concern. You should take an abstraction as a dependency and test that you only poke that dependency once (value is cached). The abstraction would look like `IInstallationIdProvider`. We could have 2 implementations (the File and the NetworkCard one) but to keep this simple it would be fine to have a single one (i.e: `DefaultInstallationIdProvider` or something that includes the code you have here). `Lazy<T>` is a good approach on that to make sure you only read that value once (only do the I/O or sys call once).

Reply:
Went with your first suggestion to keep it simpler for now